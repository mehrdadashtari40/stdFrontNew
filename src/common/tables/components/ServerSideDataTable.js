import React, { Component } from 'react';
import _ from 'lodash';
import $ from 'jquery'
class DataTablePaged extends Component {
	constructor(props) {
		super(props);
		this.state = {
			Users: [],
			total: null,
			limit: null,
			current: null,
			pageCount: null,
			startPage: null,
			endPage: null,
			startIndex: null,
			endIndex: null,
			refresh: false,
			filter: '',
		};
		this.datatable = this.datatable.bind(this);
	}

	componentDidMount() {
		this.setState({
			Users: this.props.data,
		});
		this.setState(
			{
				total: this.props.pagination.total,
				limit: this.props.pagination.limit,
				current: this.props.pagination.current != null ? this.props.pagination.current : 1,
				Users: this.props.data,
				pageCount: this.getPageCount(this.props.pagination.total, this.props.pagination.limit),
				refresh: false,
			},
			() => {
				this.calcutePageIndex();
			}
		);
	}

	setParams = object => {
		let self = this;
		return new Promise(function(resolve, reject) {
			if (!self.isEmpty(object)) {
				self.setState(object);
				resolve(true);
			} else {
				reject();
			}
		});
	};

	isEmpty = obj => {
		for (var key in obj) {
			if (obj.hasOwnProperty(key)) return false;
		}
		return true;
	};

	componentWillReceiveProps(newProps) {
		if (this.props.filter != newProps.filter) {
			let self = this;
			this.setParams({ filter: newProps.filter }).then(function(res) {
				self.getItems(10);
			});
		}

		if (newProps.refresh == true && !this.state.refresh) {
			this.setState({
				refresh: true,
			});
		} else {
			this.setState({
				refresh: false,
			});
		}
	}

	getItems = (limit, start = 0, filter = '', current) => {
		if (start == 0) this.setState({ current: start / limit + 1 });

		this.props.getDataList(
			this.props.action,
			start,
			limit,
			this.props.filter,
			this.props.sort,
			this.props.dir,
			this.props.cat_uid,
			this.props.pro_uid,
			filter,
			this.props.dateFrom,
			this.props.dateTo,
			this.props.newerThan,
			current
		);
	};

	changePaginateSize = e => {
		this.setState({
			limit: e.target.value,
		});
		this.getItems(e.target.value, 0, this.state.filter, 1);
	};

	calcutePageIndex = () => {
		if (this.state.current <= 5) {
			const pageCount = this.state.pageCount;
			let end = this.state.current + 4 < this.state.pageCount ? this.state.current + 4 : this.state.pageCount;
			this.setState({
				startIndex: 1,
				endIndex: end,
			});
		} else if (this.state.current > 5) {
			let start = this.state.current - 4;
			let end = this.state.current + 3 < this.state.pageCount ? this.state.current + 3 : this.state.pageCount;
			// this.state.pageCount;
			this.setState({
				startIndex: start,
				endIndex: end,
			});
		}
	};

	datatable() {
		const element = $(this.refs.table);
		let { options, showAct, fullTextSearch } = { ...this.props } || {};
		let toolbar = '';
		if (options.buttons) toolbar += 'B';
		if (this.props.paginationLength) toolbar += 'l';
		if (this.props.columnsHide) toolbar += 'C';

		if (typeof options.ajax === 'string') {
			let url = options.ajax;
			options.ajax = {
				url: url,
				type: 'POST',
				complete: function(xhr) {
					// AjaxActions.contentLoaded(xhr)
				},
			};
		}
		{
			/*<'col-sm-6 col-xs-12 hidden-xs text-right'" + toolbar + ">r*/
		}
		// "<'dt-toolbar-footer'<'col-sm-6 col-xs-12 hidden-xs'i><'col-xs-12 col-sm-6'p>>", // change
		// "<'dt-toolbar'<'col-xs-12 col-sm-6'f><'col-sm-6 col-xs-12 hidden-xs text-right'" + toolbar + ">r>"
		options = _.extend(options, {
			dom:
				' ' +
				't' +
				"<'dt-toolbar-footer'<'col-sm-2 col-xs-12 hidden-xs pull-right'" +
				toolbar +
				"><'col-sm-4 col-xs-12 hidden-xs pull-right'i><'col-xs-10 col-sm-6 pagination-centered'p>>",
			oLanguage: {
				sSearch: "<span class='input-group-addon input-sm'><i class='glyphicon glyphicon-search'></i></span> ",
				sEmptyTable: 'اطلاعاتی ثبت نگردیده است',
				sInfo: '_START_ تا _END_ از _TOTAL_',
				sInfoEmpty: '0 از 0',
				sInfoFiltered: '(از _MAX_ رکورد فیلتر شده)',
				sInfoPostFix: '',
				sInfoThousands: '.',
				sLengthMenu: 'نمایش _MENU_ رکورد',
				sLoadingRecords: 'در حال بارگذاری...',
				sProcessing: 'در حال پردازش...',
				sZeroRecords: 'هیچ رکوردی یافت نشد',
				oPaginate: {
					sFirst: 'اولین',
					sPrevious: 'قبلی',
					sNext: 'بعدی',
					sLast: 'آخرین',
				},
				oAria: {
					sSortAscending: ': مرتب سازی',
					sSortDescending: ': مرتب سازی',
				},
			},

			autoWidth: false,
			retrieve: true,
			responsive: true,
			destroy: true,
		});

		const _dataTable = element.DataTable(options);

		if (this.props.filterDataTable) {
			element.on('keyup change', 'thead th input[type=text]', function() {
				_dataTable
					.column(
						$(this)
							.parent()
							.index() + ':visible'
					)
					.search(this.value)
					.draw();
			});
		}

		if (this.props.showAct) {
			element.on('dblclick', 'td', function() {
				const data = _dataTable.row($(this).parents('tr')).data();
				showAct(data);
			});
		}

		if (this.props.detailsFormat) {
			const format = this.props.detailsFormat;
			element.on('click', 'td.details-control', function() {
				const tr = $(this).closest('tr');
				const row = _dataTable.row(tr);
				if (row.child.isShown()) {
					row.child.hide();
					tr.removeClass('shown');
				} else {
					row.child(format(row.data())).show();
					tr.addClass('shown');
				}
			});
		}
		if (options.rowID) {
			element.on('dblclick', 'td.row-controller', function() {
				const id = $(this)
					.siblings(options.rowID)
					.html();
				window.location.replace('#/outlook/detail/' + id);
			});
		}
	}

	render() {
		let pagination = null;
		if (this.props.data.length > 0) {
			var tmp = [];
			for (var i = this.state.startIndex; i <= this.state.endIndex; i++) tmp.push(this.getPageinateCountItem(i));

			pagination = (
				<ul className="pagination">
					<li className="paginate_button first">
						<a aria-controls="DataTables_Table_7" data-dt-idx="0" tabIndex="0" onClick={this.firstPage}>
							اولین
						</a>
					</li>
					<li className="paginate_button previous">
						<a aria-controls="DataTables_Table_7" data-dt-idx="0" tabIndex="0" onClick={this.previusPage}>
							قبلی
						</a>
					</li>
					{tmp}
					<li className="paginate_button next">
						<a aria-controls="DataTables_Table_7" data-dt-idx="6" tabIndex="0" onClick={this.nextPage}>
							بعدی
						</a>
					</li>
					<li className="paginate_button last">
						<a aria-controls="DataTables_Table_7" data-dt-idx="0" tabIndex="0" onClick={this.lastPage}>
							آخرین
						</a>
					</li>
				</ul>
			);
		}

		return (
			<div>
				<table
					className="table table-striped table-bordered table-hover dataTable no-footer dtr-inline"
					width="100%"
					filter="true"
					id="DataTables_Table_7"
					role="grid"
					aria-describedby="DataTables_Table_7_info"
				>
					{this.props.children}
				</table>

				<div className="dt-toolbar-footer">
					<div className="col-sm-4 col-xs-12 hidden-xs pull-right">
						<div className="dataTables_length" id="DataTables_Table_7_length" style={{display:"flex"}}>

							<div className="recordLabel" style={{margin:10}}>نمایش تعداد رکورد در هرصفحه</div>

							<select
								name="DataTables_Table_7_length"
								aria-controls="DataTables_Table_7"
								value={this.state.limit === null?"":this.state.limit}
								className="form-control input-sm"
								id="dataTableSize"
								onChange={this.changePaginateSize}
							>
								<option value="10">10</option>
								<option value="25">25</option>
								<option value="50">50</option>
								<option value="100">100</option>
							</select>
						</div>
					</div>
					<div className="col-xs-10 col-sm-6 pagination-centered">
						<div className="dataTables_paginate paging_simple_numbers" id="DataTables_Table_7_paginate">
							{pagination}
						</div>
					</div>
				</div>
			</div>
		);
	}

	getPageCount(total_items, limit) {
		return Math.ceil(total_items / limit);
	}

	getPageinateCountItem = i => {
		return (
			<li className={`paginate_button ${this.state.current == i ? 'active' : ''}`} key={i}>
				<a aria-controls="DataTables_Table_7" data-dt-idx="0" tabIndex="0" onClick={() => this.clickOnPage(i)}>
					{i}
				</a>
			</li>
		);
	};

	clickOnPage = index => {
		this.setState({ current: index });

		const start = this.state.limit * (index - 1);
		this.getItems(this.state.limit, start, this.props.pagination.filter, index);
	};

	previusPage = () => {
		if (this.state.current > 1) this.clickOnPage(this.state.current - 1);
	};

	nextPage = () => {
		if (this.getPageCount(this.state.total, this.state.limit) > this.state.current)
			this.clickOnPage(this.state.current + 1);
	};

	firstPage = () => {
		this.clickOnPage(1);
	};

	lastPage = () => {
		this.clickOnPage(this.getPageCount(this.state.total, this.state.limit));
	};
}

export default DataTablePaged;
