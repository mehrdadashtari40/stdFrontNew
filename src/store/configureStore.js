import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import { composeWithDevTools } from 'redux-devtools-extension';

import { userReducer, requestUserInfo } from '../common/user';
import { i18nReducer, i18nInit } from '../common/i18n';
import { trelloReducer } from '../views/trello';
import { systemManagementReducer } from '../views/system-management';
import { ProcessManagementReducer } from '../views/process-management';

import {reducer as layout2Reducer} from '../common/layout/_redux/layoutRedux'
import {reducer as tableReducer} from '../../src/common/tables/_redux/tableRedux'


import {reducer as basicInformationReducer} from '../views/basic-information/_redux/baseInfoRedux';
import LogsReducer from '../views/logs/LogReducesrs';

import BIReducer from '../views/BI/BIReducer';
import {reducer as inboxReducer} from "../views/inbox/_redux/inboxRedux";
import {reducer as authReducer} from '../../src/views/auth/_redux/authRedux'

import {reducer as ServiceDeskReducer} from "../views/service-desk/_redux/serviceDeskRedux";
import {reducer as SupervisorReducer} from "../views/supervisor/_redux/supervisorRedux";
import {reducer as CustomSettingsReducer} from "../views/customSettings/_redux/customSettingsRedux";
import {reducer as MenuReducer} from "../views/menu/_redux/menuRedux";
import {reducer as WorkDeskReducer} from "../views/work-desk/_redux/workDeskRedux";

export const rootReducer = combineReducers({
	auth: authReducer,
	user: userReducer,
	i18n: i18nReducer,
	basicInformation: basicInformationReducer,
	processManagement: ProcessManagementReducer,
	systemManagement: systemManagementReducer,
	ServiceDesk:ServiceDeskReducer,
	BI:BIReducer,
	trello:trelloReducer,
	inbox:inboxReducer,
	layout2:layout2Reducer,
	table:tableReducer,
	logs:LogsReducer,
	supervisor:SupervisorReducer,
	customSettingsReducer:CustomSettingsReducer,
	menuReducer:MenuReducer,
	workDeskReducer:WorkDeskReducer
});

const store = createStore(
	rootReducer,
	applyMiddleware(thunk)
);

store.dispatch(requestUserInfo());

store.dispatch(i18nInit());

export default store;
