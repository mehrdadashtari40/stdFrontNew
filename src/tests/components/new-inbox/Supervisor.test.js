import React from 'react';
import { shallow } from 'enzyme';
import ScalerChart from '../../../views/new-inbox/components/scaler';
import ServerSideDataTable from '../../../common/tables/components/ServerSideDataTable';
import { Provider } from 'react-redux';
import store from '../../../store/configureStore';
import configureStore from 'redux-mock-store';
import * as selectActions from '../../../views/new-inbox/NewInboxActions';
import Supervisor from '../../../views/inbox/components/Supervisor';

jest.mock('react-i18next', () => ({
	withTranslation: () => Component => {
		Component.defaultProps = { ...Component.defaultProps, t: () => '' };
		return Component;
	},
}));

test('Make Snapshot file , Supervisor Component', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	expect(wrapper).toMatchSnapshot();
});

it('invokes `componentDidMount` when mounted', () => {
	jest.spyOn(Supervisor.prototype, 'componentDidMount');
	shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	expect(Supervisor.prototype.componentDidMount).toHaveBeenCalled();
	Supervisor.prototype.componentDidMount.mockRestore();
});

test('getHistory Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getHistory).toBeDefined();
});

test('getCaseOnClicked Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getCaseOnClicked).toBeDefined();
});

test('getProccessInformation Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getProccessInformation).toBeDefined();
});

test('getTaskInformation Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getTaskInformation).toBeDefined();
});




test('getDataServer Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getDataServer).toBeDefined();
});


test('getTableRow Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Supervisor />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getTableRow).toBeDefined();
});
