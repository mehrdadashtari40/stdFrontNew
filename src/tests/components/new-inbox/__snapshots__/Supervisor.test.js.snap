// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make Snapshot file , Supervisor Component 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
      "width": "100%",
    }
  }
>
  <div
    className="col-md-12 border-box"
  >
    <div
      className="margin-top-10"
    >
      <div
        className="table-wrap custom-scroll"
      >
        <div
          className="text-center"
          style={
            Object {
              "margin": 30,
            }
          }
        >
          <b />
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="ServiceModalLabel"
    className="modal fade"
    id="caseHistoryModal"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      style={
        Object {
          "width": "65%",
        }
      }
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <button
            aria-hidden="true"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            ×
          </button>
          <h4
            className="modal-title"
            id="ServiceModalLabel"
            style={
              Object {
                "display": "inline-block",
              }
            }
          />
        </div>
        <div
          className=""
        >
          <form
            action=""
            className="smart-form"
            id="service-form"
            method="post"
          >
            <div
              className="tabbable-panel"
            >
              <div
                className="tabbable-line"
                id="tabs"
              >
                <div
                  className="tab-content add_service_modal_bx"
                >
                  <div
                    className="tab-pane active"
                    id="tab_default_4"
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn_inbox btn_datatable"
            data-dismiss="modal"
            type="button"
          >
            <i
              className="fa fa-window-close "
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="ServiceModalLabel"
    className="modal fade"
    id="processInfoModal"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      style={
        Object {
          "width": "65%",
        }
      }
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <button
            aria-hidden="true"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            ×
          </button>
          <h4
            className="modal-title"
            id="ServiceModalLabel"
            style={
              Object {
                "display": "inline-block",
              }
            }
          />
        </div>
        <div
          className=""
        >
          <form
            action=""
            className="smart-form"
            id="service-form"
            method="post"
          >
            <div
              className="tabbable-panel"
            >
              <div
                className="tabbable-line"
                id="tabs"
              >
                <div
                  className="tab-content add_service_modal_bx"
                >
                  <div
                    className="tab-pane active"
                    id="tab_default_4"
                  >
                    <Spinner />
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn_inbox btn_datatable"
            data-dismiss="modal"
            type="button"
          >
            <i
              className="fa fa-window-close "
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="ServiceModalLabel"
    className="modal fade"
    id="taskInfoModal"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      style={
        Object {
          "width": "65%",
        }
      }
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <button
            aria-hidden="true"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            ×
          </button>
          <h4
            className="modal-title"
            id="ServiceModalLabel"
            style={
              Object {
                "display": "inline-block",
              }
            }
          />
        </div>
        <div
          className=""
        >
          <form
            action=""
            className="smart-form"
            id="service-form"
            method="post"
          >
            <div
              className="tabbable-panel"
            >
              <div
                className="tabbable-line"
                id="tabs"
              >
                <div
                  className="tab-content add_service_modal_bx"
                >
                  <div
                    className="tab-pane active"
                    id="tab_default_4"
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn_inbox btn_datatable"
            data-dismiss="modal"
            type="button"
          >
            <i
              className="fa fa-window-close "
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby=""
    className="modal fade"
    id="VariableModal"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      style={
        Object {
          "width": "65%",
        }
      }
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <button
            aria-hidden="true"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            ×
          </button>
          <h4
            className="modal-title"
            id=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          />
        </div>
        <div
          className=""
        >
          <form
            action=""
            className="smart-form"
            id="service-form"
            method="post"
          >
            <div
              className="tabbable-panel"
            >
              <div
                className="tabbable-line"
                id="tabs"
              >
                <div
                  className="tab-content add_service_modal_bx"
                >
                  <table
                    className="table table-bordered"
                  >
                    <tbody>
                      <tr>
                        <td
                          className="center"
                        >
                           
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn_inbox btn_datatable"
            data-dismiss="modal"
            type="button"
          >
            <i
              className="fa fa-window-close "
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
